presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java

```diff
@@ -95,33 +119,193 @@ public abstract class IcebergAbstractMetadata
+
+        ConnectorTableLayout layout = getTableLayout(
+                session,
+                new IcebergTableLayoutHandle.Builder()
+                        .setPartitionColumns(getPartitionKeyColumnHandles(icebergTable, typeManager))
+                        .setDomainPredicate(constraint.getSummary().transform(IcebergAbstractMetadata::toSubfield))
+                        .setRemainingPredicate(TRUE_CONSTANT)
+                        .setPredicateColumns(predicateColumns)
+                        .setRequestedColumns(requestedColumns)
+                        .setPushdownFilterEnabled(isPushdownFilterEnabled(session))
+                        .setPartitionColumnPredicate(partitionColumnPredicate)
+                        .setPartitions(Optional.empty())
+                        .setTable(handle)
+                        .build());
         return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
     }
 
+    public static Subfield toSubfield(ColumnHandle columnHandle)
+    {
+        return new Subfield(((IcebergColumnHandle) columnHandle).getName(), ImmutableList.of());
+    }
+
+    protected static boolean isEntireColumn(Subfield subfield)
+    {
+        return subfield.getPath().isEmpty();
+    }
+
     @Override
     public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle)
     {
-        return new ConnectorTableLayout(handle);
+        if (!isPushdownFilterEnabled(session)) {
+            return new ConnectorTableLayout(handle);
+        }
+
+        IcebergTableLayoutHandle icebergTableLayoutHandle = (IcebergTableLayoutHandle) handle;
+
+        IcebergTableHandle tableHandle = icebergTableLayoutHandle.getTable();
+        if (!tableExists(session, tableHandle.getSchemaTableName())) {
+            return null;
+        }
+
+        Table icebergTable = getIcebergTable(session, tableHandle.getSchemaTableName());
+
+        List<ColumnHandle> partitionColumns = ImmutableList.copyOf(icebergTableLayoutHandle.getPartitionColumns());
+        List<HivePartition> partitions = icebergTableLayoutHandle.getPartitions().get();
+
+        Optional<DiscretePredicates> discretePredicates = Optional.empty();
+        if (!partitionColumns.isEmpty()) {
+            // Do not create tuple domains for every partition at the same time!
+            // There can be a huge number of partitions so use an iterable so
+            // all domains do not need to be in memory at the same time.
+            Iterable<TupleDomain<ColumnHandle>> partitionDomains = Iterables.transform(partitions, (hivePartition) -> TupleDomain.fromFixedValues(hivePartition.getKeys()));
+            discretePredicates = Optional.of(new DiscretePredicates(partitionColumns, partitionDomains));
+        }
+
+        TupleDomain<ColumnHandle> predicate;
+        RowExpression subfieldPredicate;
+        if (icebergTableLayoutHandle.isPushdownFilterEnabled()) {
```

Comment:
You were using isPushdownFilterEnabled(session) on line 194. Can they be different? If not, can you please use one same way to check this?